web/auth/twofactor.go

```diff
@@ -0,0 +1,147 @@
+package auth
+
+import (
+	"net/http"
+	"net/url"
+	"strconv"
+
+	"github.com/cozy/cozy-stack/model/instance"
+	"github.com/cozy/cozy-stack/model/oauth"
+	"github.com/cozy/cozy-stack/pkg/limits"
+	"github.com/cozy/cozy-stack/web/middlewares"
+	"github.com/cozy/echo"
+)
+
+const (
+	TrustedDeviceCheckBoxActivated   = true
+	TrustedDeviceCheckBoxDeActivated = false
+)
+
+func renderTwoFactorForm(c echo.Context, i *instance.Instance, code int, credsError string, redirect *url.URL, twoFactorToken []byte, longRunSession bool, trustedDeviceCheckBox bool) error {
+	title := i.Translate("Login Two factor title")
+
+	redirectQuery := redirect.Query()
+	var clientScope string
+	if clientScopes := redirectQuery["scope"]; len(clientScopes) > 0 {
+		clientScope = clientScopes[0]
+	}
+
+	oauth := i.HasDomain(redirect.Host) && redirect.Path == "/auth/authorize" && clientScope != oauth.ScopeLogin
+	return c.Render(code, "twofactor.html", echo.Map{
+		"CozyUI":                middlewares.CozyUI(i),
+		"ThemeCSS":              middlewares.ThemeCSS(i),
+		"Domain":                i.ContextualDomain(),
+		"ContextName":           i.ContextName,
+		"Locale":                i.Locale,
+		"Title":                 title,
+		"CredentialsError":      credsError,
+		"Redirect":              redirect.String(),
+		"LongRunSession":        longRunSession,
+		"TwoFactorToken":        string(twoFactorToken),
+		"CSRF":                  c.Get("csrf"),
+		"OAuth":                 oauth,
+		"Favicon":               middlewares.Favicon(i),
+		"TrustedDeviceCheckBox": trustedDeviceCheckBox,
+	})
+}
+
+// twoFactorForm handles the twoFactor from GET request
+func twoFactorForm(c echo.Context) error {
+	var credsError string
+	trustedDeviceCheckBox := true
+
+	inst := middlewares.GetInstance(c)
+
+	redirect, err := checkRedirectParam(c, inst.DefaultRedirection())
+	if err != nil {
+		return err
+	}
+
+	twoFactorTokenParam := c.QueryParams().Get("two_factor_token")
+	if twoFactorTokenParam == "" {
+		return c.JSON(http.StatusBadRequest, "Missing twoFactorToken")
+	}
+
+	twoFactorToken := []byte(twoFactorTokenParam)
+
+	longRunSession, err := strconv.ParseBool(c.FormValue("long-run-session"))
+	if err != nil {
+		longRunSession = true
+	}
+
+	trustedDeviceCheckBoxParam := c.QueryParams().Get("trusted_device_checkbox")
+	if trustedDeviceCheckBoxParam != "" {
+		if b, err := strconv.ParseBool(trustedDeviceCheckBoxParam); err == nil {
+			trustedDeviceCheckBox = b
+		}
+	}
+
+	return renderTwoFactorForm(c, inst, http.StatusOK, credsError, redirect, twoFactorToken, longRunSession, trustedDeviceCheckBox)
+}
+
+// twoFactor handles a the twoFactor POST request
+func twoFactor(c echo.Context) error {
```

Comment:
This function is a bit too long to my taste. Can you extract an helper? Maybe a func to handle the 2fa failure